### PR TITLE
Frontpage + Bump feature

### LINKED ISSUE #18: Move discovery to frontpage
Image

### LINKED ISSUE #19: Display a list of active rants on frontpage
`frontend/api.js` exports publicKernel.

Display publicKernel.$rants() as list.

### LINKED ISSUE #20: As a user I want the rants to be sorted by weight
In general there are 2 types of rants.

1. Uncurated / fresh new rants, short lived.
2. Scored / Curated long-lived rants

### LINKED ISSUE #23: Create a 'dope' button on rant
When clicked:
```js
await pubKernel.bump(rantId)
```

### LINKED ISSUE #24: pubKernel.bump(rant)
The filter:
https://github.com/decentlabs-north/rant/blob/cc65c01229170f16c33b995469033acaeebb6640/blockend/public-kernel.js#L93

Filter logic:
Reject if already bumped 10 times.
Reject if bump.date < max(rant.date, rant.lastBumpedAt)
Reject if bump.date > Date.now() // in future.

Picture:

Image

### LINKED ISSUE #25: As a user i want to be able to choose if my rant is going to be public or private

### PR DESCRIPTION
**Pull request by Pingu1337**
- #18 
- #19 
- #20 
- #23 
- #24

### CODE REVIEW

File: blockend/public-kernel.js [hunk 1]
```diff
@@ -66,6 +68,33 @@ export default class PublicKernel extends SimpleKernel {
...
+    }
+    await this.createBlock(branch, TYPE_BUMP, data)
+
+    // branch.inspect() <-- debugging purposes
+  }
+
+  /**
+   * This is used to get the clients key before any actions have been made.
+   * There is probably a better and less hacky way to do it.
+   */
```
Comment (telamon) [inline blockend/public-kernel.js:89]: +points for hacky solution. there's `this.pk` would that work?
https://github.com/telamon/picostack/blob/master/simple-kernel.js#L39

File: blockend/public-kernel.js [hunk 2]
```diff
@@ -89,9 +118,27 @@ function TinyBoard (size = 50, ttl = ONE_HOUR, now = Date.now) {
...
         } break
 
-        case TYPE_BUMP:
+        case TYPE_BUMP:{
           // TODO: model stun-lock with diminishing returns
-          break
+          const rant = state[btok(CHAIN)] // rant
+
+          // get who is bumping
+          const hasBumped = rant.bumpedBy.some((key) => btok(key) === btok(block.key))
```
Comment (telamon) [inline blockend/public-kernel.js:126]: if you want minor speed perf improvement you can use buffer.equals(otherbuffer): `key.equals(block.key)`.
logic is legit

File: blockend/public-kernel.js [hunk 3]
```diff
@@ -89,9 +118,27 @@ function TinyBoard (size = 50, ttl = ONE_HOUR, now = Date.now) {
...
+          if (hasBumped) return 'AlreadyBumped'
+
+          if (data.date > now()) return 'TimeManipulation'
+
+          if (data.date > rant.expiresAt) return 'TooLate'
+
+          if (data.date < rant.lastBumpAt) return 'TimeManipulation'
+
+          const timeSinceLastBump = (now() - rant.lastBumpAt)
+          if (timeSinceLastBump < 1000) return 'TooSoonToBump'
```
Comment (telamon) [inline blockend/public-kernel.js:136]: Nice, I didn't think of this, maybe even raise the number to 10-30seconds to give the new chain state some more time to propagate through the network. it's a very useful mechanic to slow down bump-spam/unwarranted bumps.

File: frontend/components/discover-page.js [hunk 4]
```diff
@@ -6,6 +6,11 @@ import { processText, THEMES } from '../../blockend/picocard.js'
 import Modem56 from 'picostack/modem56.js'
 import dayjs from '../day.js'
 
+/**
+ * THIS FILE WILL BE DELETED.
+ * IM ONLY KEEPING IT IN CASE I FUDGE SOMETHING UP ON THE FRONTPAGE-
```
Comment (telamon) [inline frontend/components/discover-page.js:11]: 👍  delete at will.

File: frontend/components/frontpage-feed.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        return this.html`
+          <rant-card rant=${rant} key=${this.key}></rant-card>
+        `
+      })
+    }
+
+    return this.html`
+    <div style="text-align: center;">
+      <h1 style="font-size: 6rem; margin-bottom:0.25rem;">1k</h1>
+      <p>Tin Foil Twitter</p>
```
Comment (telamon) [inline frontend/components/frontpage-feed.js:88]: mixed like. don't like using other brands but easy to relate. How about _Real Times_ news network? 😆

**Review by telamon — APPROVED** [review]
dope! very so. See if `this.pk` can be used instead of `this.$key` then merge.

File: blockend/public-kernel.js (hunk 1 above)
Comment (Pingu1337) [reply blockend/public-kernel.js:89]: thx was looking for this x)

File: blockend/public-kernel.js (hunk 3 above)
Comment (Pingu1337) [reply blockend/public-kernel.js:136]: 10sec should be a good number

File: frontend/components/frontpage-feed.js (hunk 5 above)
Comment (Pingu1337) [reply frontend/components/frontpage-feed.js:88]: This was only memes, you pick the slogan :smile_cat:

**Comment by Pingu1337** [comment]
> dope! very so. See if `this.pk` can be used instead of `this.$key` then merge.
@telamon there is one small issue at the moment that i would like to resolve before merging.
Client written posts are still not public. I will take a look and see if i can find a solution!

**Pull request merged**